### DISCUSSION TITLE
Chain loading config

### DISCUSSION THREAD

**Discussion by KarelWintersky** [PHPAuth/PHPAuth#541] [Ideas]
My plan:

```php

// init PDO before all things

$config = new Config($pdo);

// here: initializing default config with
// - default locale
// - default config
// - no validators 

// load* methods load and override previous loaded (or default) config:

$config = $config->loadDB(); // .. from default phpauth_config table (set this at installation stage)
// or
$config = $config->loadDB($pdo_connection, $config_table); // ... from config_table via other $pdo_connection
// or
$config = $config->loadDB(null, $config_table); // ... from same connection and other table)

// or (Array)

$custom_settings = [
    'cookie_name' => getenv('PHPAUTH.COOKIE'),
    'site_email' => getenv('PHPAUTH.SIME_MAIL'),
    'site_url' => getenv('PHPAUTH.SITE_URL)
];

$config = $config->loadArray($config_dataset); // override previously loaded settings with data from array

// or (I used this method only once):

$config = $config->loadINI( '/etc/livemap/phpauth.ini' );

// using loadXML, loadYAML, loadJSON is questionable, can be discussed. adding them requires additional libraries or extensions.

$config = $config->setLocalization( (new \PHPAuth\PHPAuthLocalization('fr_FR'))->use() );

// define custom validators:

$config = $config->setPasswordValidator(static function($password) use ($config) {
    return (bool)((new Zxcvbn())->passwordStrength($password)['score'] >= intval($config->password_min_score));
});

$config = $config->setEMailValidator(static function ($email) {
    return \PHPAuth\EMailValidator::isValid($email);
});

$config = $config->setCaptchaValidator(static function() use ($reCaptcha_config) {
    if (empty($reCaptcha_config)) {
        return true;
    }

    if ($reCaptcha_config['recaptcha_enabled']) {
        if (empty($reCaptcha_config['recaptcha_secret_key'])) {
            throw new RuntimeException('No secret provided');
        }

        if (!is_string($reCaptcha_config['recaptcha_secret_key'])) {
            throw new RuntimeException('The provided secret must be a string');
        }

        $recaptcha = new ReCaptcha($reCaptcha_config['recaptcha_secret_key']);
        $checkout = $recaptcha->verify($captcha_response, \PHPAuth\Helpers::getIp());

        if (!$checkout->isSuccess()) {
            return false;
        }
    }

    return true;
}, $reCaptcha_config);

// now, config is ready, instantiate Auth:

$auth = new Auth($pdo, $config->ready() );

// put $auth to Application Container and so on... 

```

### necessary and sufficient code for work:

```php

// init PDO before all things

$config = new Config($pdo);

$config = $config->loadArray([
    'site_name' => getenv('PHPAUTH.SITE_NAME'),
    'site_email' => getenv('PHPAUTH.SITE_MAIL'),
    'site_url' => getenv('PHPAUTH.SITE_URL),
    'site_timezone' => getenv('PHPAUTH.SITE_TIMEZONE'),
    'cookie_name' => getenv('PHPAUTH.COOKIE'),
    'uses_session' => 1 // for example
]); 

$config = $config->setLocalization( (new \PHPAuth\PHPAuthLocalization('fr_FR'))->use() );

$auth = new Auth($pdo, $config->ready() );
```

that's all. 

I think such code is much more transparent, understandable and extensible than the current architecture.

Default settings, only partially overridden by options that _really_ need to be changed.

The default language is English, downloadable via a language pack.

A password validator, optionally included - and it can be not only `bjeavons/zxcvbn-php`, but any other validator that returns true or false.

Email validator - similar.

Captcha validator - through a callback - not only google reCaptcha, but also Yandex, and kCaptcha and any other self-written one.

Similar to mail handler.

**Comment by KarelWintersky** [comment]
```
/**
     * Load config from XML.
     *
     * Beginner `$` in filepath means application root directory. Same effect for JSON, YAML and XML files
     *
     * Required `ext-simplexml` and `ext-libxml` in composer.json
     *
     * @param null $filename
     * @return $this
     * @throws PHPAuthException
     */
    public function loadXML($filename = null):self
    {
        if (!function_exists('simplexml_load_file')) {
            throw new PHPAuthException( 'PHPAuth Config: simplexml_load_file() function not implemented' );
        }

        if (is_null($filename)) {
            throw new PHPAuthException( 'PHPAuth Config: given empty XML config file name' );
        }

        $filename = self::resolveAppRootPath($filename);

        if (!is_file($filename)) {
            throw new PHPAuthException("PHPAuth Config: XML-file {$filename} does not exist.");
        }

        if (!is_readable($filename)) {
            throw new PHPAuthException( "PHPAuth Config: XML-file {$filename} is unreadable." );
        }

        $content = \file_get_contents($filename);
        if (false === $content) {
            throw new PHPAuthException( 'PHPAuth Config: Can\' get file content' );
        }

        libxml_use_internal_errors(true); // don't display xml errors

        $xml = simplexml_load_string($content); // 'ext-simplexml' is missing in composer.json

        if (false === $xml) {
            throw new PHPAuthException( 'PHPAuth Config: XML file is corrupted' );
        }

        $this->config->set($xml);

        return $this;
    }
```

**Comment by KarelWintersky** [comment]
```
 /**
     * Load config from YAML. Using built-in function yaml_parse
     * Beginner `$` in filepath means application root directory. Same effect for JSON, YAML and XML files
     *
     * https://www.php.net/manual/ru/function.yaml-parse.php (required yaml extension)
     * required at least ext-yaml extension
     *
     * @param null $filename
     * @return $this
     * @throws PHPAuthException
     *
     * see also: https://symfony.com/doc/current/components/yaml.html :
     * use Symfony\Component\Yaml\Yaml;
     * $value = Yaml::parseFile('/path/to/file.yaml');
     * 
     */
    public function loadYAML($filename = null):self
    {
        if (!function_exists('yaml_parse')) {
            throw new PHPAuthException( 'PHPAuth Config: yaml_parse() function not implemented' );
        }

        if (is_null($filename)) {
            throw new PHPAuthException( 'PHPAuth Config: given empty YAML config file name' );
        }

        $filename = self::resolveAppRootPath($filename);

        // replace beginner '$' in filepath to application root directory
        $filename = preg_replace('/^\$/', getcwd(), $filename);

        if (!is_file($filename)) {
            throw new PHPAuthException("PHPAuth Config: YAML-file {$filename} does not exist.");
        }

        if (!is_readable($filename)) {
            throw new PHPAuthException( "PHPAuth Config: YAML-file {$filename} is unreadable." );
        }

        $content = \file_get_contents($filename);
        if (false === $content) {
            throw new PHPAuthException( 'PHPAuth Config: Can\' get file content' );
        }

        $yaml = yaml_parse($content); // Note: by default, 'ext-yaml' is missing in composer.json

        if (false === $yaml)
            throw new PHPAuthException('PHPAuth Config: YAML file is corrupted');

        $this->config->set($yaml);

        return $this;
    }
```

**Comment by KarelWintersky** [comment]
```
    /**
     * Load PHPAuth Config from JSON-file
     * Beginner `$` in filepath means application root directory.
     *
     * @param null $filename
     * @return $this
     * @throws PHPAuthException
     */
    public function loadJSON($filename = null):self
    {
        if (is_null($filename)) {
            throw new PHPAuthException( 'PHPAuth Config: can\'t load configuration from undefined file!' );
        }

        $filename = self::resolveAppRootPath($filename);

        if (!is_file($filename)) {
            throw new PHPAuthException("PHPAuth Config: JSON-file {$filename} does not exist.");
        }

        if (!is_readable($filename)) {
            throw new PHPAuthException( "PHPAuth Config: JSON-file {$filename} is unreadable." );
        }

        $content = \file_get_contents($filename);
        if (false === $content) {
            throw new PHPAuthException( "PHPAuth Config: Can't get {$filename} content");
        }

        $json = json_decode($content, true);

        // https://www.php.net/manual/ru/function.json-last-error-msg.php
        if (json_last_error() !== JSON_ERROR_NONE) {
            throw new PHPAuthException( 'PHPAuth Config: JSON is corrupted. Error ['.json_last_error().']: '.json_last_error_msg() );
        }

        $this->config->set($json);

        return $this;
    }
```

**Comment by KarelWintersky** [comment]
```
    /**
     * Load PHPAuth Config from INI-file
     * Beginner `$` in filepath means application root directory.
     *
     * @param $filename
     * @return $this
     * @throws PHPAuthException
     */
    public function loadINI($source = null):self
    {
        if (empty($source)) {
            throw new PHPAuthException( 'PHPAuth Config: can\'t load configuration from undefined file!' );
        }

        $source = self::resolveAppRootPath($source);

        if (!is_file($source)) {
            throw new PHPAuthException("PHPAuth Config: INI-file {$source} does not exist.");
        }

        if (!is_readable($source)) {
            throw new PHPAuthException( "PHPAuth Config: INI-file {$source} is unreadable." );
        }

        $ini = parse_ini_file($source, true, INI_SCANNER_TYPED);

        if (false === $ini) {
            throw new PHPAuthException( 'PHPAuth Config: INI file is corrupted' );
        }

        // load configuration
        $this->config->set($ini);

        $this->config_type = self::CONFIG_TYPE_INI;

        return $this;
    }
```

**Comment by Maikuolan** [comment]
> Captcha validator - through a callback - not only google reCaptcha, but also Yandex, and kCaptcha and any other self-written one.

Multiple options is definitely a good idea. reCaptcha is super popular, enjoys market dominance nowadays, and most users will probably want to use that. So, in regard to CAPTCHAs, supporting reCaptcha is definitely important. But, due to concerns about privacy, targeted advertising, and potential tracking and spying (regardless of whether these things actually happen, there is still a significant minority of users which distrust Google, and will have these concerns), some users might be unhappy if reCaptcha was the only option.

Considering those concerns, [HCaptcha](https://www.hcaptcha.com/) could be worth investigating, too. Their motto: "Private. Secure. Faster." Claims to be built with privacy in mind, without any tracking, spying, etc. Most websites/packages/etc which I've built and maintain provide support for reCaptcha and HCaptcha.

More options (e.g., Yandex, kCaptcha, etc) is even better. But of course, more options also means more work, so it also depends on how much work you want to commit to doing. ;-)

> // or (I used this method only once):
>
> $config = $config->loadINI( '/etc/livemap/phpauth.ini' );
>
> // using loadXML, loadYAML, loadJSON is questionable, can be discussed. adding them requires additional libraries or extensions.

Generally speaking, INI files are okay to use. Most people will be familiar with how INI files work, they're simple, easy to use, etc.

The main problems with INI files is that (1) custom data types can't be natively supported, and (2) they're limited to two-dimensional data. But, it doesn't matter much, because if really needed, you can easily fake both things by just serialising the data first, INI interprets it as a string, and when deserialised by the implementation, have the implementation perform whatever processing could be needed, or for data which is than two dimensions (matrices, complex objects, etc), use dot notation, delimit data in a similar way to CSV data, etc.

Like you said, some formats will require additional libraries or extensions. That may be undesirable for some users. Of course, if a particular format is strongly desired, luckily, there are many such libraries and extensions available which could be implemented.

For YAML, if we wanted something reasonably small and simple (some libraries have multiple sub-dependencies, swathes of directories, files, etc), there's a small library of classes I wrote a while back which contains a YAML handler ([**link here**](https://github.com/Maikuolan/Common)). You're welcome to use that, if you want. Doesn't need any extensions. ;-)

It's what I use for configuration files, L10N files, and various other data for a number of my own projects. </Shameless Plug> 👀

But, of course, supporting everything isn't necessary. Just a few things, like a PDO connection, INI files, etc, should be perfectly okay, IMO.

> Default settings, only partially overridden by options that *really* need to be changed.

I do the exact same thing for most of my packages nowadays, so.. I definitely agree with that strategy. ^^

**Comment by KarelWintersky** [comment]
> But of course, more options also means more work, so it also depends on how much work you want to commit to doing. ;-)

And I don't plan to write a custom handler for every possible captcha. I have developed and want to implement a mechanism in the core that allows you to use any arbitrary captcha through a callback.

The callback function passed to the PHPAuth\Config should figure out the result of passing the captcha itself, possibly by getting data from the arguments (to be honest, we did not test this mechanism in detail).

As a small bonus, I will give examples of a couple of options for using various captchas.

```php
$config = $config->setCaptchaValidator(static function() use ($reCaptcha_config) {
    if (empty($reCaptcha_config)) {
        return true;
    }

    if ($reCaptcha_config['recaptcha_enabled']) {
        if (empty($reCaptcha_config['recaptcha_secret_key'])) {
            throw new RuntimeException('No secret provided');
        }

        if (!is_string($reCaptcha_config['recaptcha_secret_key'])) {
            throw new RuntimeException('The provided secret must be a string');
        }

        $recaptcha = new ReCaptcha($reCaptcha_config['recaptcha_secret_key']);
        $checkout = $recaptcha->verify($captcha_response, \PHPAuth\Helpers::getIp());

        if (!$checkout->isSuccess()) {
            return false;
        }
    }

    return true;
}, $reCaptcha_config);
```

By the way, due to some technical/political limitations, reCaptcha may not be available in some countries.

**Comment by Maikuolan** [comment]
LGTM. :+1:

**Comment by KarelWintersky** [comment]
> I wrote a while back which contains a YAML handler ([link here](https://github.com/Maikuolan/Common)). You're welcome to use that, if you want.

thank you. 
can i use your code to create a package PHPAuth\Yaml-Parser and lay it out via compositor for use via suggest section ?

**Comment by Maikuolan** [comment]
> thank you.
> can i use your code to create a package PHPAuth\Yaml-Parser and lay it out via compositor for use via suggest section ?

Yes, you can do that. The licensing I applied to the overall package (GNU/GPLv2) already permits copying and redistribution anyway (as long as the users of the redistributed copy can inherit the same right to copy and redistribute), so no problem. :-)

(Originally chose GNU/GPLv2 in order to match with some other packages I had written previously, but have contemplated maybe switching to MIT eventually, due to MIT being a little more permissive and less verbose. Anyway, in any case, no problem).
